usgpo/get-links.py

```diff
@@ -0,0 +1,185 @@
+
+
+def get_packages(api_key, collections, start_date, package_queue):
+    logger = logs.get_logger("usgpo")
+    url = f"https://api.govinfo.gov/published/{start_date}"
+    offset_mark = "*"
+    pbar = tqdm(desc="Producer")
+    while url is not None:
+        response = api_query(
+            url,
+            headers={"accept": "application/json"},
+            params={
+                "api_key": args.api_key,
+                "offsetMark": offset_mark,
+                "pageSize": 1000,
+                "collection": ",".join(collections),
+            },
+        )
+        if response.status_code == 200:
+            output = response.json()
+
+            for record in output["packages"]:
+                package_queue.put(record)
+                pbar.update(1)
+
+            url = output["nextPage"]
+            offset_mark = None
+            # Prevent too many API requests in a short period of time
+            time.sleep(5)
+        else:
+            logger.error(
+                f"get_packages received status code {response.status_code} for query {url}"
+            )
+            break
+
+    package_queue.put(None)
+
+
+def get_file_links(api_key, package):
+    package_id = package["packageId"]
+    response = api_query(
+        f"https://api.govinfo.gov/packages/{package_id}/summary",
+        headers={"accept": "application/json"},
+        params={"api_key": args.api_key},
+    )
+    if response.status_code == 200:
+        output = response.json()
+        return output.get("download")
+    return None
+
+
+def get_package_metadata(api_key, package_queue, metadata_queue):
+    pbar = tqdm(desc="Consumer")
+    while True:
+        package = package_queue.get()
+        if package is None:
+            package_queue.put(None)
+            metadata_queue.put(None)
+            break
+
+        record = {
+            "title": package.get("title"),
+            "package_id": package.get("packageId"),
+            "date": package.get("dateIssued"),
+            "category": package.get("category"),
+            "author": package.get("governmentAuthor1"),
+            "publisher": package.get("publisher"),
+            "links": get_file_links(api_key, package),
+        }
+        metadata_queue.put(record)
+        pbar.update(1)
+
+
+def write_metadata(output_dir, metadata_queue):
+    with jsonlines.open(os.path.join(output_dir, "links.jsonl"), mode="w") as writer:
+        pbar = tqdm(desc="Writer")
+        while True:
+            metadata = metadata_queue.get()
+            if metadata is None:
+                metadata_queue.task_done()
+                break
+
+            writer.write(metadata)
+            pbar.update(1)
+
+
+def main(args):
+    os.makedirs(args.output_dir, exist_ok=True)
+
+    package_queue = queue.Queue()
+    metadata_queue = queue.Queue()
+
+    with ThreadPoolExecutor(max_workers=args.workers + 2) as executor:
```

Comment:
This parallelism is pretty complicated, it is possible to fetch all the packages in parallel, collect results, and then fetch the metadata from each package in parallel afterwards?